File: gdx/src/com/badlogic/gdx/graphics/g3d/utils/TextureProvider.java
package com.badlogic.gdx.graphics.g3d.utils;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.assets.AssetManager;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g3d.Model;
import com.badlogic.gdx.graphics.g3d.model.data.ModelData;

/**
 * Used by {@link Model} to load textures from {@link ModelData}.
 *
 */
public interface TextureProvider {
    Texture load(String fileName);

    class FileTextureProvider implements TextureProvider {
        private final Texture.TextureFilter minFilter;
        private final Texture.TextureFilter magFilter;
        private final Texture.TextureWrap uWrap;
        private final Texture.TextureWrap vWrap;
        private final boolean useMipMaps;

        public FileTextureProvider() {
            minFilter = magFilter = Texture.TextureFilter.Linear;
            uWrap = vWrap = Texture.TextureWrap.Repeat;
            useMipMaps = false;
        }

        public FileTextureProvider(Texture.TextureFilter minFilter, Texture.TextureFilter magFilter, Texture.TextureWrap uWrap,
                                   Texture.TextureWrap vWrap, boolean useMipMaps) {
            this.minFilter = minFilter;
            this.magFilter = magFilter;
            this.uWrap = uWrap;
            this.vWrap = vWrap;
            this.useMipMaps = useMipMaps;
        }

        @Override
        public Texture load(String fileName) {
            Texture result = new Texture(Gdx.files.internal(fileName), useMipMaps);
            result.setFilter(minFilter, magFilter);
            result.setWrap(uWrap, vWrap);
            return result;
        }
    }

    class AssetTextureProvider implements TextureProvider {
        public final AssetManager assetManager;

        public AssetTextureProvider(final AssetManager assetManager) {
            this.assetManager = assetManager;
        }

        @Override
        public Texture load(String fileName) {
            return assetManager.get(fileName, Texture.class);
        }
    }
}

File: gdx/src/com/badlogic/gdx/graphics/g3d/utils/TextureProvider.kt
package com.badlogic.gdx.graphics.g3d.utils

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.assets.AssetManager
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.Texture.TextureFilter
import com.badlogic.gdx.graphics.Texture.TextureWrap
import com.badlogic.gdx.graphics.g3d.Model
import com.badlogic.gdx.graphics.g3d.model.data.ModelData

/**
 * Used by [Model] to load textures from [ModelData].
 */
interface TextureProvider {

    fun load(fileName: String): Texture

    class FileTextureProvider @JvmOverloads constructor(
        private val minFilter: TextureFilter = TextureFilter.Linear,
        private val magFilter: TextureFilter = TextureFilter.Linear,
        private val uWrap: TextureWrap = TextureWrap.Repeat,
        private val vWrap: TextureWrap = TextureWrap.Repeat,
        private val useMipMaps: Boolean = false
    ) : TextureProvider {

        override fun load(fileName: String): Texture {
            val result = Texture(Gdx.files.internal(fileName), useMipMaps)
            result.setFilter(minFilter, magFilter)
            result.setWrap(uWrap, vWrap)
            return result
        }
    }

    class AssetTextureProvider(val assetManager: AssetManager) : TextureProvider {
        override fun load(fileName: String): Texture = assetManager.get(fileName, Texture::class.java)
    }
}
